automata/regex/parser.py

```diff
@@ -216,6 +216,39 @@ def option(self):
         self._initial_state = new_initial_state
         self._final_states.add(new_initial_state)
 
+    def shuffle(self, other):
+        """
+        Apply the shuffle operation to the NFA represented by this builder and other.
+        No need for BFS since all states are acessible.
+        """
+        new_state_name_dict = dict()
+
+        def get_state_name(state_name):
+            return new_state_name_dict.setdefault(state_name, self.__get_next_state_name())
+
+        self._initial_state = get_state_name((self._initial_state, other._initial_state))
+
+        new_transitions = dict()
+
+        for (q_a, transitions_a), (q_b, transitions_b) in product(
+            self._transitions.items(), other._transitions.items()
+        ):
```

Comment:
@eliotwrobson I might suggest creating a dedicated variable for the result of the `product(...)` so the `for` loop definition can fit on one line. Something like:

```python
transition_product = product(self._transitions.items(), other._transitions.items())
for (q_a, transitions_a), (q_b, transitions_b) in transition_product:
    # ...
```